File: app/src/main/java/piuk/blockchain/android/ui/pairing_code/PairingCodePresenter.java
package piuk.blockchain.android.ui.pairing_code;

import android.graphics.Bitmap;

import javax.inject.Inject;

import io.reactivex.Observable;
import okhttp3.ResponseBody;
import piuk.blockchain.android.R;
import piuk.blockchain.android.data.datamanagers.AuthDataManager;
import piuk.blockchain.android.data.datamanagers.PayloadDataManager;
import piuk.blockchain.android.data.datamanagers.QrCodeDataManager;
import piuk.blockchain.android.data.rxjava.RxUtil;
import piuk.blockchain.android.injection.Injector;
import piuk.blockchain.android.ui.base.BasePresenter;
import piuk.blockchain.android.ui.customviews.ToastCustom;
import piuk.blockchain.android.util.StringUtils;

public class PairingCodePresenter extends BasePresenter<PairingCodeView> {

    private static final String WEB_WALLET_URL = "blockchain.info/wallet/login";

    @Inject QrCodeDataManager qrCodeDataManager;
    @Inject StringUtils stringUtils;
    @Inject PayloadDataManager payloadDataManager;
    @Inject AuthDataManager authDataManager;

    public PairingCodePresenter() {
        Injector.getInstance().getDataManagerComponent().inject(this);
    }

    @Override
    public void onViewReady() {
        //no op
    }

    String getFirstStep() {
        return String.format(stringUtils.getString(R.string.pairing_code_instruction_1), WEB_WALLET_URL);
    }

    public void generatePairingQr() {
        getPairingEncryptionPasswordObservable()
                .doOnSubscribe(disposable -> getView().showProgressSpinner())
                .doAfterTerminate(() -> getView().hideProgressSpinner())
                .flatMap(encryptionPassword -> generatePairingCodeObservable(encryptionPassword.string()))
                .compose(RxUtil.addObservableToCompositeDisposable(this))
                .subscribe(
                        bitmap -> getView().onQrLoaded(bitmap),
                        throwable -> getView().showToast(R.string.unexpected_error, ToastCustom.TYPE_ERROR));
    }

    private Observable<ResponseBody> getPairingEncryptionPasswordObservable() {

        String guid = payloadDataManager.getWallet().getGuid();
        return authDataManager.getPairingEncryptionPassword(guid);
    }

    private Observable<Bitmap> generatePairingCodeObservable(String encryptionPhrase) {

        String guid = payloadDataManager.getWallet().getGuid();
        String sharedKey = payloadDataManager.getWallet().getSharedKey();
        String password = payloadDataManager.getTempPassword();

        return qrCodeDataManager.generatePairingCode(
                guid,
                password,
                sharedKey,
                encryptionPhrase,
                600);
    }
}

File: app/src/main/java/piuk/blockchain/android/ui/pairing_code/PairingCodePresenter.kt
package piuk.blockchain.android.ui.pairing_code

import android.graphics.Bitmap
import io.reactivex.Observable
import okhttp3.ResponseBody
import piuk.blockchain.android.R
import piuk.blockchain.android.data.datamanagers.AuthDataManager
import piuk.blockchain.android.data.datamanagers.PayloadDataManager
import piuk.blockchain.android.data.datamanagers.QrCodeDataManager
import piuk.blockchain.android.data.rxjava.RxUtil
import piuk.blockchain.android.ui.base.BasePresenter
import piuk.blockchain.android.ui.customviews.ToastCustom
import piuk.blockchain.android.util.StringUtils
import javax.inject.Inject

class PairingCodePresenter @Inject constructor(
        private val qrCodeDataManager: QrCodeDataManager,
        stringUtils: StringUtils,
        private val payloadDataManager: PayloadDataManager,
        private val authDataManager: AuthDataManager
) : BasePresenter<PairingCodeView>() {

    override fun onViewReady() {
        // No op
    }

    internal val firstStep =
            String.format(stringUtils.getString(R.string.pairing_code_instruction_1), WEB_WALLET_URL)

    fun generatePairingQr() {
        pairingEncryptionPasswordObservable
                .doOnSubscribe { view.showProgressSpinner() }
                .doAfterTerminate { view.hideProgressSpinner() }
                .flatMap { encryptionPassword -> generatePairingCodeObservable(encryptionPassword.string()) }
                .compose(RxUtil.addObservableToCompositeDisposable<Bitmap>(this))
                .subscribe(
                        { bitmap -> view.onQrLoaded(bitmap) },
                        { view.showToast(R.string.unexpected_error, ToastCustom.TYPE_ERROR) })
    }

    private val pairingEncryptionPasswordObservable: Observable<ResponseBody>
        get() {
            val guid = payloadDataManager.wallet.guid
            return authDataManager.getPairingEncryptionPassword(guid)
        }

    private fun generatePairingCodeObservable(encryptionPhrase: String): Observable<Bitmap> {
        val guid = payloadDataManager.wallet.guid
        val sharedKey = payloadDataManager.wallet.sharedKey
        val password = payloadDataManager.tempPassword

        return qrCodeDataManager.generatePairingCode(
                guid,
                password,
                sharedKey,
                encryptionPhrase,
                600)
    }

    companion object {
        private val WEB_WALLET_URL = "blockchain.info/wallet/login"
    }
}
